### PR TITLE
Add systematic blocking

### PR DESCRIPTION
**Pull request by mikemahoney218**
This PR adds a systematic method for assigning blocks to folds, numbering blocks in a grid sequentially starting from the bottom-left corner of the grid. If `ordering = "snake"`, the default, this ordering moves left-to-right for the first row of grid cells, right-to-left for the second row of grid cells, and continues the pattern from there. If `ordering = "continuous"`, ordering is left-to-right for each row. If `relevant_only = TRUE`, only blocks containing data are labelled; if `relevant_only = FALSE` all blocks are labeled, potentially resulting in entire folds with no data.

After this, we've implemented 3/3 block assignment methods in [blockCV](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13107) (one of them via `spatial_clustering_cv()`, by doing eg `spatial_clustering_cv(ames, coords = c(Lot_Area))`) and 3/4 methods in [Roberts et al 2016](https://onlinelibrary.wiley.com/doi/full/10.1111/ecog.02881). 

The missing method from Roberts is an "optimized random" sampling, where folds are specifically assigned to minimize dissimilarity in predictor space. I'd appreciate talking with you all about this -- they mention it as a way to assess interpolation accuracy when a model won't be used to extrapolate, but I'm concerned about supporting that and potentially undermining the "pit of success" we've got here by making it too easy to assess your model on a purposefully easy benchmark.

(ETA: We're agreed that we're not going to move forward on incorporating predictor-space into blocking at the moment)

``` r
library(spatialsample)
library(ggplot2)
data(ames, package = "modeldata")
data(Sacramento, package = "modeldata")

ames_sf <- sf::st_as_sf(ames, coords = c("Longitude", "Latitude"), crs = 4326)

Sacramento_sf <- sf::st_as_sf(Sacramento, coords = c("longitude", "latitude"), crs = 4326)

make_plot <- function(data, title) {
  do.call(
    rbind, 
    lapply(
      seq_len(nrow(data)), 
      \(x) cbind(assessment(data$splits[[x]]), fold = factor(x))
    )
  ) |> 
    ggplot(aes(color = fold)) + 
    geom_sf() + 
    coord_sf(crs = 4326) + 
    theme_minimal() + 
    labs(title = title)
}

set.seed(123)
spatial_block_cv(ames_sf, v = 5, square = FALSE, cellsize = 0.005, method = "systematic") |> 
  make_plot(title = "Ames 5-fold")
```

``` r
spatial_block_cv(ames_sf, 
         v = 5, 
         square = FALSE, 
         n = 100, 
         method = "systematic",
         ordering = "continuous",
         relevant_only = FALSE) |> 
  make_plot(title = "Ames 5-fold")
```

``` r
spatial_block_cv(Sacramento_sf, v = 20, method = "systematic") |> 
  make_plot(title = "Sacramento 20-fold")
```

<sup>Created on 2022-05-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

**Review by hfrick — APPROVED** [review]
Looks good! I left a suggestion about improving how the error messages are shown to the user, which you could apply to lower-level check functions. If you use that, it'd be good to require rlang minimum version 1.0.0.

**Comment by mikemahoney218** [comment]
Seems to be some test failures related to the RStudio CDN. Not sure what to do about that other than to come back later.

**Comment by mikemahoney218** [comment]
Looks like the outage is wrapping up -- CI passed at last.

File: R/spatial_block_cv.R [hunk 1]
```diff
@@ -104,54 +118,67 @@ spatial_block_cv <- function(data, method = "random", v = 10, ...) {
...
+  if (num_folds != v) {
+    rlang::warn(
+      c(
+        paste0(
+          "Not all folds contained blocks with data: \n",
+          v,
+          " folds were requested, but only ",
+          num_folds,
+          " contain any data. \nEmpty folds were dropped."
+        ),
```
Comment (juliasilge) [inline R/spatial_block_cv.R:168]: Let's switch over to `glue::glue()` for our error messages in this package, starting here. You can add glue to Imports.

File: R/spatial_block_cv.R [hunk 2]
```diff
@@ -170,3 +197,56 @@ random_block_cv <- function(data, grid_blocks, v) {
...
+filter_grid_blocks <- function(grid_blocks, data) {
+  block_contains_points <- purrr::map_lgl(
+    sf::st_intersects(grid_blocks, data),
+    sgbp_is_not_empty
+  )
+  if ("data.frame" %in% class(grid_blocks)) {
+    grid_blocks[block_contains_points, ]
+  } else {
+    grid_blocks[block_contains_points]
+  }
```
Comment (juliasilge) [inline R/spatial_block_cv.R:210]: Do [we need `drop = FALSE` here](https://adv-r.hadley.nz/subsetting.html#simplify-preserve)?

File: R/spatial_block_cv.R [hunk 3]
```diff
@@ -170,3 +197,56 @@ random_block_cv <- function(data, grid_blocks, v) {
...
+row_ids_intersecting_fold_blocks <- function(grid_blocks, data) {
+  # grid_blocks is a list of sgbp lists (?sf::sgbp)
+  #
+  # The first map() here iterates through the meta-list,
+  # and the second checks each element of the relevant sgbp list
+  # to see if it is integer(0) (no intersections) or not
+  #
+  # Each sgbp sub-list is nrow(data) elements long, so this which()
+  # returns the list indices which are not empty, which is equivalent
+  # to the row numbers that intersect with blocks in the fold
```
Comment (juliasilge) [inline R/spatial_block_cv.R:242]: This is so helpful! 🙏

File: R/spatial_clustering_cv.R [hunk 4]
```diff
@@ -122,8 +120,8 @@ spatial_clustering_splits <- function(data, dists, v = 10, cluster_function = c(
   indices <- split_unnamed(idx, folds)
   indices <- lapply(indices, default_complement, n = n)
   split_objs <- purrr::map(indices, make_splits,
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:122]: Make this a new line break (after `map(`) and then reindent this function call.

File: tests/testthat/test-spatial_clustering_cv.R [hunk 5]
```diff
@@ -83,9 +83,9 @@ test_that("bad args", {
 test_that("can pass the dots to kmeans", {
   expect_error(
     spatial_clustering_cv(Smithsonian,
```
Comment (juliasilge) [inline tests/testthat/test-spatial_clustering_cv.R:85]: I'd break the line after `spatial_cluster_cv(` and then reindent

**Review by juliasilge — CHANGES_REQUESTED** [review]
I think we need either a new approach or more error checking for the arguments:

``` r
library(spatialsample)
data(Smithsonian, package = "modeldata")
smithsonian_sf <- sf::st_as_sf(Smithsonian,
                               coords = c("longitude", "latitude"),
                               # Set CRS to WGS84
                               crs = 4326)
set.seed(123)
spatial_block_cv(smithsonian_sf, v = 3)
#> #  3-fold spatial block cross-validation 
#> # A tibble: 3 × 2
#>   splits         id   
#>   <list>         <chr>
#> 1 <split [3/17]> Fold1
#> 2 <split [18/2]> Fold2
#> 3 <split [19/1]> Fold3
set.seed(123)
spatial_block_cv(smithsonian_sf, v = 3, method = "systematic")
#> #  3-fold spatial block cross-validation 
#> # A tibble: 3 × 2
#>   splits         id   
#>   <list>         <chr>
#> 1 <split [19/1]> Fold1
#> 2 <split [3/17]> Fold2
#> 3 <split [18/2]> Fold3
set.seed(123)
spatial_block_cv(smithsonian_sf, v = 3, method = "systematic", ordering = "continuous")
#> #  3-fold spatial block cross-validation 
#> # A tibble: 3 × 2
#>   splits         id   
#>   <list>         <chr>
#> 1 <split [19/1]> Fold1
#> 2 <split [3/17]> Fold2
#> 3 <split [18/2]> Fold3
set.seed(123)
spatial_block_cv(smithsonian_sf, v = 3, method = "random", ordering = "continuous")
#> #  3-fold spatial block cross-validation 
#> # A tibble: 3 × 2
#>   splits         id   
#>   <list>         <chr>
#> 1 <split [3/17]> Fold1
#> 2 <split [18/2]> Fold2
#> 3 <split [19/1]> Fold3
```

<sup>Created on 2022-05-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

In the last case, the user will get `method = "random"` and passing in `ordering = "continuous"` doesn't do anything but also does not error.

I'm trying to find other examples we have of arguments that only are applied when a previous argument has a certain value. Do you have some good patterns of this you have seen?

One idea is to make an argument like `systematic_opts = list()`. You can document what goes in the list, add an example, check the input (that is it NULL if "random" and the names if "systematic"), and keep `rlang::arg_match()` in the lower level function for checking `ordering = c("snake", "continuous")`. I'm not saying definitely do this, but it is one option for making this more foolproof.

**Comment by hfrick** [comment]
Awesome! 🎉 Just named the arg, otherwise good to go @juliasilge !

**Pull request merged**